modules/evm/src/lib.rs

```diff
@@ -890,18 +909,26 @@ impl<T: Config> Pallet<T> {
 				}
 			});
 
-			AccountStorages::<T>::remove_prefix(address, None);
-
-			let size = ContractStorageSizes::<T>::take(address);
-
-			Ok(size)
+			T::IdleScheduler::schedule(
+				EvmTask::Remove {
+					caller: *caller,
+					contract: *address,
+					maintainer: contract_info.maintainer,
+				}
+				.into(),
+			)
 		})?;
 
 		// this should happen after `Accounts` is updated because this could trigger another updates on
 		// `Accounts`
 		frame_system::Pallet::<T>::dec_providers(&address_account)?;
 
-		Ok(size)
+		// estimate weight
+		let weight = (ContractStorageSizes::<T>::get(address)
```

Comment:
I don't think it is useful to charge additional weight here. The tx fee will be charged based on max gas and refund unused but it cannot charge more if more weights are returned.

The additional removal work is done on idle block time, so it is ok for them to be free. We may also only refund x% of the deposit.